File: src/ui/constellation_view.cpp
#include "ui/constellation_view.h"

#include <imgui.h>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstring>

namespace jalert::ui {

ConstellationView::ConstellationView(ConstellationPull pull)
    : pull_(std::move(pull)) {
    scratch_.resize(kPullPerFrame);
}

void ConstellationView::reset() {
    std::memset(accum_, 0, sizeof(accum_));
    last_accum_max_ = 0.0f;
    observed_amp_ = 1.0;
}

void ConstellationView::decay_accumulator() {
    for (int i = 0; i < kBinCount * kBinCount; ++i) accum_[i] *= kAccumDecay;
}

void ConstellationView::accumulate_and_stats(int n) {
    samples_this_frame_ = n;
    clipped_this_frame_ = 0;

    const float ideal_axis = static_cast<float>(observed_amp_);  // BPSK: |Re|≈amp
    double evm_sq = 0.0, amp_sq = 0.0;
    int evm_n = 0;

    for (int i = 0; i < n; ++i) {
        const dsp::Complex32 z = scratch_[i];
        const float re = z.re, im = z.im;

        const float ideal_re = re >= 0.0f ? ideal_axis : -ideal_axis;
        const float d_re = re - ideal_re;
        const float d_im = im;                          // ideal Im = 0
        evm_sq += d_re * d_re + d_im * d_im;
        amp_sq += re * re + im * im;
        ++evm_n;

        if (std::fabs(re) > kAmpMax || std::fabs(im) > kAmpMax) {
            ++clipped_this_frame_;
            continue;
        }
        const int bx = static_cast<int>((re + kAmpMax) * (kBinCount / (2.0f * kAmpMax)));
        const int by = static_cast<int>((kAmpMax - im) * (kBinCount / (2.0f * kAmpMax)));
        if (static_cast<unsigned>(bx) >= kBinCount ||
            static_cast<unsigned>(by) >= kBinCount) continue;
        accum_[by * kBinCount + bx] += 1.0f;
    }

    if (evm_n > 0) {
        const double frame_amp = std::sqrt(amp_sq / evm_n);
        if (frame_amp > 0.05) {
            observed_amp_ = (1.0 - kObservedAmpAlpha) * observed_amp_
                            + kObservedAmpAlpha * frame_amp;
        }
        const double denom = std::max(0.1, observed_amp_);
        evm_percent_ = std::sqrt(evm_sq / evm_n) / denom * 100.0;
    } else {
        evm_percent_ = 0.0;
    }
}

void ConstellationView::draw(float edge_px) {
    if (edge_px < 16.0f) edge_px = 16.0f;

    ImDrawList* dl = ImGui::GetWindowDrawList();
    const ImVec2 cur = ImGui::GetCursorScreenPos();
    const float ox = cur.x, oy = cur.y, size = edge_px;

    dl->AddRectFilled(ImVec2(ox, oy), ImVec2(ox + size, oy + size),
                      IM_COL32(0, 0, 0, 255));

    const int pull_n = pull_ ? pull_(scratch_.data(),
                                     static_cast<int>(scratch_.size())) : 0;
    decay_accumulator();
    accumulate_and_stats(pull_n);

    const float cx = ox + size * 0.5f;
    const float cy = oy + size * 0.5f;
    const float scale = (size * 0.5f) / kAmpMax;

    const ImU32 dim    = IM_COL32(50, 220, 50, 40);
    const ImU32 marker = IM_COL32(50, 220, 50, 110);

    dl->AddRect(ImVec2(ox, oy), ImVec2(ox + size - 1, oy + size - 1), dim, 0.0f, 0, 1.0f);
    dl->AddLine(ImVec2(ox, cy), ImVec2(ox + size, cy), dim, 1.0f);
    dl->AddLine(ImVec2(cx, oy), ImVec2(cx, oy + size), dim, 1.0f);

    const float orbit = std::max(0.1f, std::min(kAmpMax, static_cast<float>(observed_amp_)));
    const float ideal_r = scale * orbit;
    const float m = std::max(3.0f, size / 50.0f);
    auto cross = [&](float px, float py) {
        dl->AddLine(ImVec2(px - m, py - m), ImVec2(px + m, py + m), marker, 1.0f);
        dl->AddLine(ImVec2(px - m, py + m), ImVec2(px + m, py - m), marker, 1.0f);
    };
    cross(cx + ideal_r, cy);    // BPSK ideal points on the real axis
    cross(cx - ideal_r, cy);

    float max_v = 0.0f;
    const float bin_px = size / static_cast<float>(kBinCount);
    constexpr float kAlphaScale = 32.0f;
    for (int i = 0; i < kBinCount * kBinCount; ++i) {
        const float v = accum_[i];
        if (v > max_v) max_v = v;
        int a = static_cast<int>(v * kAlphaScale);
        if (a <= 0) continue;
        if (a > 255) a = 255;
        const int by = i / kBinCount, bx = i % kBinCount;
        const float x0 = ox + bx * bin_px, y0 = oy + by * bin_px;
        dl->AddRectFilled(ImVec2(x0, y0), ImVec2(x0 + bin_px + 1.0f, y0 + bin_px + 1.0f),
                          IM_COL32(50, 220, 50, a));
    }
    last_accum_max_ = max_v;

    const ImU32 dot_col = IM_COL32(50, 220, 50, 220);
    const int take = std::min(pull_n, kDotsOverlayCount);
    const float l = ox, t = oy, r = ox + size, b_ = oy + size;
    for (int i = 0; i < take; ++i) {
        const dsp::Complex32 z = scratch_[i];
        const float px = cx + z.re * scale;
        const float py = cy - z.im * scale;
        if (px < l || px >= r || py < t || py >= b_) continue;
        dl->AddRectFilled(ImVec2(px - 1.0f, py - 1.0f), ImVec2(px + 1.0f, py + 1.0f), dot_col);
    }

    if (pull_n > 0) {
        char buf[32];
        std::snprintf(buf, sizeof(buf), "EVM %.1f%%", evm_percent_);
        const ImVec2 sz = ImGui::CalcTextSize(buf);
        dl->AddText(ImVec2(ox + size - sz.x - 4.0f, oy + 2.0f),
                    IM_COL32(50, 220, 50, 190), buf);

        if (samples_this_frame_ > 0 &&
            static_cast<double>(clipped_this_frame_) / samples_this_frame_ >= kClipWarnRatio) {
            const double pct = 100.0 * clipped_this_frame_ / samples_this_frame_;
            char cbuf[40];
            std::snprintf(cbuf, sizeof(cbuf), "! clip %.0f%%", pct);
            const ImVec2 csz = ImGui::CalcTextSize(cbuf);
            dl->AddText(ImVec2(ox + size - csz.x - 4.0f, oy + size - csz.y - 2.0f),
                        IM_COL32(255, 140, 0, 220), cbuf);
        }
    }

    if (pull_n == 0 && last_accum_max_ < 0.05f) {
        const char* msg = "no signal";
        const ImVec2 sz = ImGui::CalcTextSize(msg);
        dl->AddText(ImVec2(cx - sz.x * 0.5f, cy - sz.y * 0.5f),
                    IM_COL32(50, 220, 50, 140), msg);
    }

    ImGui::Dummy(ImVec2(size, size));
}

} // namespace jalert::ui

File: SDRSharp.JAlert/Ui/ConstellationView.cs
// Real-time scatter / heatmap of the Costas-derotated BPSK symbols, rendered with
// GDI+. BPSK clusters at (±amp, 0), so EVM and the ideal markers reference the
// real axis. Ported from src/ui/constellation_view.{h,cpp} (ImGui → WinForms).

using System;
using System.Drawing;
using System.Windows.Forms;
using SDRSharp.JAlert.Dsp;

namespace SDRSharp.JAlert.Ui
{
    // Newest-first pull (matches BpskDemod.PullConstellation()).
    public delegate int ConstellationPull(Complex32[] dst, int capacity);

    public sealed class ConstellationView : Control
    {
        private const int BinCount = 80;
        private const float AccumDecay = 0.88f;
        private const int PullPerFrame = 256;
        private const int DotsOverlayCount = 96;
        private const float AmpMax = 1.5f;
        private const double ClipWarnRatio = 0.02;
        private const double ObservedAmpAlpha = 0.05;

        private static readonly Color Green = Color.FromArgb(50, 220, 50);

        private ConstellationPull _pull;
        private readonly Complex32[] _scratch = new Complex32[PullPerFrame];
        private readonly float[] _accum = new float[BinCount * BinCount];

        private double _evmPercent;
        private double _evmIPercent;   // decision-axis (Re) error — drives BER
        private double _evmQPercent;   // quadrature (Im) error — phase/benign
        private double _snrDb;         // Es/N0 estimate from the decision axis
        private int _clippedThisFrame;
        private int _samplesThisFrame;
        private float _lastAccumMax;
        private double _observedAmp = 1.0;

        public ConstellationView()
        {
            DoubleBuffered = true;
            SetStyle(ControlStyles.ResizeRedraw | ControlStyles.UserPaint |
                     ControlStyles.AllPaintingInWmPaint | ControlStyles.OptimizedDoubleBuffer, true);
            BackColor = Color.Black;
        }

        public void SetPull(ConstellationPull pull) => _pull = pull;

        public void ResetView()
        {
            Array.Clear(_accum, 0, _accum.Length);
            _lastAccumMax = 0.0f;
            _observedAmp = 1.0;
        }

        private void DecayAccumulator()
        {
            for (int i = 0; i < _accum.Length; ++i) _accum[i] *= AccumDecay;
        }

        private void AccumulateAndStats(int n)
        {
            _samplesThisFrame = n;
            _clippedThisFrame = 0;

            float idealAxis = (float)_observedAmp;   // BPSK: |Re|≈amp
            double evmReSq = 0.0, evmImSq = 0.0, ampSq = 0.0;
            int evmN = 0;

            for (int i = 0; i < n; ++i)
            {
                float re = _scratch[i].Re, im = _scratch[i].Im;

                float idealRe = re >= 0.0f ? idealAxis : -idealAxis;
                float dRe = re - idealRe;
                float dIm = im;                          // ideal Im = 0
                evmReSq += dRe * dRe;                     // decision-axis error
                evmImSq += dIm * dIm;                     // quadrature error
                ampSq += re * re + im * im;
                ++evmN;

                if (Math.Abs(re) > AmpMax || Math.Abs(im) > AmpMax)
                {
                    ++_clippedThisFrame;
                    continue;
                }
                int bx = (int)((re + AmpMax) * (BinCount / (2.0f * AmpMax)));
                int by = (int)((AmpMax - im) * (BinCount / (2.0f * AmpMax)));
                if ((uint)bx >= BinCount || (uint)by >= BinCount) continue;
                _accum[by * BinCount + bx] += 1.0f;
            }

            if (evmN > 0)
            {
                double frameAmp = Math.Sqrt(ampSq / evmN);
                if (frameAmp > 0.05)
                    _observedAmp = (1.0 - ObservedAmpAlpha) * _observedAmp + ObservedAmpAlpha * frameAmp;
                double denom = Math.Max(0.1, _observedAmp);
                double sigmaI = Math.Sqrt(evmReSq / evmN);
                double sigmaQ = Math.Sqrt(evmImSq / evmN);
                _evmPercent = Math.Sqrt((evmReSq + evmImSq) / evmN) / denom * 100.0;
                _evmIPercent = sigmaI / denom * 100.0;
                _evmQPercent = sigmaQ / denom * 100.0;
                // Decision-axis Es/N0: signal amp vs the noise std on Re. This is
                // what predicts BPSK BER (the quadrature spread does not).
                _snrDb = sigmaI > 1e-6 ? 20.0 * Math.Log10(denom / sigmaI) : 99.0;
            }
            else
            {
                _evmPercent = _evmIPercent = _evmQPercent = 0.0;
                _snrDb = 0.0;
            }
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics g = e.Graphics;
            int side = Math.Max(16, Math.Min(Width, Height));
            float size = side;
            float ox = 0, oy = 0;

            g.Clear(Color.Black);

            int pullN = _pull != null ? _pull(_scratch, _scratch.Length) : 0;
            DecayAccumulator();
            AccumulateAndStats(pullN);

            float cx = ox + size * 0.5f;
            float cy = oy + size * 0.5f;
            float scale = (size * 0.5f) / AmpMax;

            using (Pen dim = new Pen(Color.FromArgb(40, Green)))
            {
                g.DrawRectangle(dim, ox, oy, size - 1, size - 1);
                g.DrawLine(dim, ox, cy, ox + size, cy);
                g.DrawLine(dim, cx, oy, cx, oy + size);
            }

            float orbit = Math.Max(0.1f, Math.Min(AmpMax, (float)_observedAmp));
            float idealR = scale * orbit;
            float m = Math.Max(3.0f, size / 50.0f);
            using (Pen marker = new Pen(Color.FromArgb(110, Green)))
            {
                DrawCross(g, marker, cx + idealR, cy, m);  // BPSK ideal points on the real axis
                DrawCross(g, marker, cx - idealR, cy, m);
            }

            float maxV = 0.0f;
            float binPx = size / BinCount;
            const float alphaScale = 32.0f;
            for (int i = 0; i < _accum.Length; ++i)
            {
                float v = _accum[i];
                if (v > maxV) maxV = v;
                int a = (int)(v * alphaScale);
                if (a <= 0) continue;
                if (a > 255) a = 255;
                int by = i / BinCount, bx = i % BinCount;
                float x0 = ox + bx * binPx, y0 = oy + by * binPx;
                using (SolidBrush br = new SolidBrush(Color.FromArgb(a, Green)))
                    g.FillRectangle(br, x0, y0, binPx + 1.0f, binPx + 1.0f);
            }
            _lastAccumMax = maxV;

            int take = Math.Min(pullN, DotsOverlayCount);
            float l = ox, t = oy, r = ox + size, b = oy + size;
            using (SolidBrush dot = new SolidBrush(Color.FromArgb(220, Green)))
            {
                for (int i = 0; i < take; ++i)
                {
                    float px = cx + _scratch[i].Re * scale;
                    float py = cy - _scratch[i].Im * scale;
                    if (px < l || px >= r || py < t || py >= b) continue;
                    g.FillRectangle(dot, px - 1.0f, py - 1.0f, 2.0f, 2.0f);
                }
            }

            using (Font font = new Font(FontFamily.GenericSansSerif, 7.0f))
            {
                if (pullN > 0)
                {
                    string buf = "EVM " + _evmPercent.ToString("0.0") + "%";
                    SizeF sz = g.MeasureString(buf, font);
                    using (SolidBrush br = new SolidBrush(Color.FromArgb(190, Green)))
                        g.DrawString(buf, font, br, ox + size - sz.Width - 4.0f, oy + 2.0f);

                    // I/Q split + decision-axis SNR: tells whether the EVM is
                    // decision-relevant (I) or benign phase/quadrature (Q).
                    string buf2 = "I " + _evmIPercent.ToString("0") + " Q " + _evmQPercent.ToString("0") +
                                  "  " + _snrDb.ToString("0.0") + "dB";
                    SizeF sz2 = g.MeasureString(buf2, font);
                    using (SolidBrush br = new SolidBrush(Color.FromArgb(150, Green)))
                        g.DrawString(buf2, font, br, ox + size - sz2.Width - 4.0f, oy + 2.0f + sz.Height);

                    if (_samplesThisFrame > 0 &&
                        (double)_clippedThisFrame / _samplesThisFrame >= ClipWarnRatio)
                    {
                        double pct = 100.0 * _clippedThisFrame / _samplesThisFrame;
                        string cbuf = "! clip " + pct.ToString("0") + "%";
                        SizeF csz = g.MeasureString(cbuf, font);
                        using (SolidBrush br = new SolidBrush(Color.FromArgb(220, 255, 140, 0)))
                            g.DrawString(cbuf, font, br, ox + size - csz.Width - 4.0f, oy + size - csz.Height - 2.0f);
                    }
                }

                if (pullN == 0 && _lastAccumMax < 0.05f)
                {
                    string msg = "no signal";
                    SizeF sz = g.MeasureString(msg, font);
                    using (SolidBrush br = new SolidBrush(Color.FromArgb(140, Green)))
                        g.DrawString(msg, font, br, cx - sz.Width * 0.5f, cy - sz.Height * 0.5f);
                }
            }
        }

        private static void DrawCross(Graphics g, Pen p, float px, float py, float m)
        {
            g.DrawLine(p, px - m, py - m, px + m, py + m);
            g.DrawLine(p, px - m, py + m, px + m, py - m);
        }
    }
}
